internal/nodevalidator/cni_detector.go

```diff
@@ -0,0 +1,260 @@
+package nodevalidator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	cniConfigDir = "/etc/cni/net.d"
+	cniBinDir    = "/opt/cni/bin"
+)
+
+// cniDetector implements CNIDetector interface
+type cniDetector struct {
+	client    kubernetes.Interface
+	logger    *zap.Logger
+	configDir string
+	binDir    string
+}
+
+// NewCNIDetector creates a new CNIDetector
+func NewCNIDetector(client kubernetes.Interface, logger *zap.Logger) CNIDetector {
+	return &cniDetector{
+		client:    client,
+		logger:    logger,
+		configDir: cniConfigDir,
+		binDir:    cniBinDir,
+	}
+}
+
+// DetectCNI checks if a supported CNI plugin is installed and running
+func (cd *cniDetector) DetectCNI(ctx context.Context, nodeName string) (CNIType, error) {
+	var cniType CNIType
+	var err error
+
+	// Validate Static CNI files on node
+	cniType, err = cd.detectStaticCNI()
+	if err != nil {
+		return cniType, err
+	}
+
+	// Validate CNI running on node object from the cluster
+	if nodeName != "" {
+		node, err := cd.client.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+		if err != nil {
+			return CNITypeNone, fmt.Errorf("could not get node object for CNI detection: %w", err)
+		}
+
+		// Check node.Status.Conditions for NetworkUnavailable condition. Most reliable method set by CNI plugin.
+		cniType, err = cd.detectFromNodeCondition(node)
+		if err == nil && cniType != CNITypeNone {
+			return cniType, nil
+		}
+
+		// Check if taint applied (node.Spec.Taints), validate only if node runtime condition failed
+		cniType = cd.detectFromNodeTaintsStatus(node)
+		if cniType != CNITypeNone {
+			return cniType, nil
+		}
+	}
+
+	return CNITypeNone, errors.New("Unable to detect CNI in Kubelet NetworkUnavailable condition")
+}
+
+// detectStaticCNI performs static validation by checking config and binaries files
+func (cd *cniDetector) detectStaticCNI() (CNIType, error) {
```

Comment:
Can we change this to just ensure the static files exist? We're not using the cniType returned by this function anywhere so we should be good just validating that it exists.